source/slang/slang-preprocessor.cpp

```diff
@@ -222,6 +222,9 @@ struct Preprocessor
         /// The module, if any, that the preprocessed result will belong to
     Module*                                 parentModule = nullptr;
 
+        /// The AST builder that should be used when creating AST nodes for `parentModule`
```

Comment:
I agree that in principle it would be cleaner to have the preprocessor not interact with the AST this much. It is also unfortunate to have the preprocessor have a baked-in understanding of which `#define`s are semantically relevant.

A weak defense is that the `Module` (and hence the `ModuleDecl`) was already being passed down into the preprocessor, so the coupling was already there.

There's basically two ways I can see to not infect the preprocessor with this much AST-level policy:

1. Have the preprocessor produce an additional output that is a dictionary of "semantically relevant" macros

2. Have the preprocessor's public API include multiple phases, and a query API, so that code driving the preprocessor can look up macros after other preprocessing is done

I initially wrote code along the lines of (1), but found that it was actually more complicated to have this logic split between the preprocessor (which would know which macros are semantically relevant) and the higher level coordination logic (which would know what AST nodes to create). Moving everything down into the preprocessor itself resulted in simpler code (and a more clear way to emit not-entirely-terrible diagnostics).

Doing something along the lines of (2) is probably a better long-term option, but would require a rather larger refactor of the current preprocessor API. The current preprocessor code keeps the actual `Preprocessor` type private to the `.cpp` file, and doesn't require it to be heap-allocated. The obvious approach for (2) would involve moving the "guts" of the preprocessor (all its data structures) out into the `.h` file, which I'd like to avoid.

One option would be to define a kind of callback interface that the client passes down into the preprocessor, which gives them a chance to insert custom logic at various points along the way. That would allow the front-end of the compiler to inject the NVAPI-specific logic as a callback that gets invoked at the end of the source unit, but before the preprocessor has been cleaned up.

Do you have thoughts on what you'd prefer to see here?

Reply:
The way you went with your subsequent PR seems to clean up things pretty well. 

If I'd put my pedant hat on I might not call it a Handler. It seems more like it's a Listener. It doesn't 'Handle' any of the preprocessor logic. It's there to be informed when certain things happen. Although Callback doesn't really stake a claim of being Handler/Listener like, and so could be used in perhaps either - I personally reserve the name for C style callbacks. Visitor doesn't fit the bill either, because that implies the traversing of a structure and the pattern from the GOF.

Anyhoo not super important, just a thought.